dlib/te.h

```diff
@@ -0,0 +1,440 @@
+            }
+        };
+
+        struct storage_heap
+        {
+            storage_heap() = default;
+
+            template <
+                class T,
+                class T_ = std::decay_t<T>,
+                std::enable_if_t<!std::is_same<T_,storage_heap>::value, bool> = true
+            >
+            storage_heap(T &&t) noexcept(std::is_nothrow_constructible<T_,T&&>::value)
+            :   ptr{new T_{std::forward<T>(t)}},
+                del{[](void *self) { 
+                    delete reinterpret_cast<T_*>(self); 
+                }},
+                copy{[](const void *self) -> void * {
+                    return new T_{*reinterpret_cast<const T_*>(self)};
+                }},
+                type_id{[] {
+                    return std::type_index{typeid(T_)};
+                }}
+            {
+            }
+
+            template <
+                class T,
+                class T_ = std::decay_t<T>,
+                std::enable_if_t<!std::is_same<T_,storage_heap>::value, bool> = true
+            >
+            storage_heap& operator=(T &&t) noexcept(std::is_nothrow_constructible<T_,T&&>::value)
+            {
+                if (contains<T_>())
+                    unsafe_get<T_>() = std::forward<T>(t);
+                else
+                    *this = std::move(storage_heap{std::forward<T>(t)});
+                return *this;
+            }
+
+            storage_heap(const storage_heap& other)
+            :   ptr{other.ptr ? other.copy(other.ptr) : nullptr},
+                del{other.del},
+                copy{other.copy},
+                type_id{other.type_id}
+            {
+            }
+
+            storage_heap& operator=(const storage_heap& other)
+            {
+                if (this != &other) 
+                    *this = std::move(storage_heap{other});
+                return *this;
+            }
+
+            storage_heap(storage_heap&& other) noexcept
+            :   ptr{std::exchange(other.ptr, nullptr)},
+                del{std::exchange(other.del, nullptr)},
+                copy{std::exchange(other.copy, nullptr)},
+                type_id{std::exchange(other.type_id, nullptr)}
+            {
+            }
+
+            storage_heap& operator=(storage_heap&& other) noexcept
+            {
+                if (this != &other) 
+                {
+                    storage_heap{std::move(*this)};
+                    ptr     = std::exchange(other.ptr, nullptr);
+                    del     = std::exchange(other.del, nullptr);
+                    copy    = std::exchange(other.copy, nullptr);
+                    type_id = std::exchange(other.type_id, nullptr);
+                }
+                return *this;
+            }
+
+            ~storage_heap()
+            {
+                if (ptr)
+                    del(ptr);
+            }
+
+            void clear()
+            {
+                storage_heap{std::move(*this)};
+            }
+
+            bool is_empty() const
```

Comment:
Maybe we need some kind of `storage_base` base class that implements `is_empty()`, `contains<T>()`, `unsafe_get<T>()`. Maybe that would remove boiler plate. Will try later